lib/SILOptimizer/Analysis/EscapeAnalysis.cpp

```diff
@@ -2157,10 +2158,20 @@ void EscapeAnalysis::analyzeInstruction(SILInstruction *I,
         return isPointer(result);
       }));
       return;
-
+    case SILInstructionKind::BeginBorrowInst:
+    case SILInstructionKind::CopyValueInst: {
+      auto SVI = cast<SingleValueInstruction>(I);
+      if (CGNode *PointsTo = ConGraph->getValueContent(SVI->getOperand(0))) {
```

Comment:
I'm pretty sure this is not correct. Both instructions are simple forwarding instructions and should be handled with `ConGraph->defer`

Reply:
@eeckstein @atrick Can you tell me/point to documentation on when to use `ConGraph->setNode` vs `ConGraph->defer` ?

Reply:
There are three ways we can relate SILValues to each other in escape analysis

1. getPointerBase

When the SILValue is directly derived from its operand and all lookups can be redirected to the operand. We don't want to distinguish between them at all for EscapeAnalysis.

2. setNode(Value2, Node1)

This maps multiple values to the same node.

The only happens when an instruction refers to a node deeper in the connection graph than its operand. For example, two ref_element_addr's with the same object root map to the same node for the object header. And two loads of the same address get mapped to the same node.

This doesn't really makes sense for copy_value, since the operand has the *same* node as the result. If we want to treat them the same, then we should just use getPointerBase.

3. defer(Node2, Node1)

This merges the content of the two nodes. It does not merge the two nodes themselves. So, in theory they can still have different escape information. In practice, I we rarely make use of that.

For OSSA lifetime and borrow scope optimization, we need to be able to ask whether a particular OSSA reference escapes. e.g. does this borrowed reference escape its borrow scope (via a store or pointer cast). If not, then we can freely shrink the borrow scope regardless of any other aliasing or escaping information!

So, we definitely want to represent a copy_value and its operand with separate nodes, and different escape information. Although in practice it may not make any difference to optimization with way the current escape analysis queries are structured. More work is probably needed. In the meantime, I still think it's the best choice.

Reply:
Thanks! Updated the PR.